### PR TITLE
Update xgboost_synthetic test infra; preliminary updates to work with 0.7.0

### PR DESCRIPTION
**Pull request by jlewi**
Update xgboost_synthetic test infra to use pytest and pyfunc.

* Related to #655 update xgboost_synthetic for 0.7

* Related to to #665 no signal about xgboost_synthetic

* We need to update the xgboost_synthetic example to work with 0.7.0;
  e.g. workload identity

* This PR focuses on updating the test infra and some preliminary
  updates the notebook

* More fixes to the test and the notebook are probably needed in order
  to get it to actually pass

* Update job spec for 0.7; remove the secret and set the default service
  account.

  * This is to make it work with workload identity

* Instead of using kustomize to define the job to run the notebook we can just modify the YAML spec using python.
* Use the python API for K8s to create the job rather than shelling out.

* Notebook should do a 0.7 compatible check for credentials

  * We don't want to assume GOOGLE_APPLICATION_CREDENTIALS is set
    because we will be using workload identity.

* Take in repos as an argument akin to what checkout_repos.sh requires

* Convert xgboost_test.py to a pytest.

  * This allows us to mark it as expected to fail so we can start to get
    signal without blocking

  * We also need to emit junit files to show up in test grid.

* Convert the jsonnet workflow for the E2E test to a python function to
  define the workflow.
  
  * Remove the old jsonnet workflow.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/examples/666)

### CODE REVIEW

File: xgboost_synthetic/testing/xgboost_test.py [hunk 1]
```diff
@@ -1,100 +1,125 @@
...
+  # TODO(jlewi): We should just assume that kubeconfig has been set.
+  if cluster:
+    util.run(['gcloud', '--project=kubeflow-ci-deployment', 'container',
+              "clusters", "get-credentials", "--zone=us-east1-b", cluster],
+               cwd=app_dir)
+
+  with open("job.yaml") as hf:
+    job = yaml.load(hf)
+
+  job["metadata"]["namespace"] = name
```
[the remark this answers is not in the document]
Comment (jlewi) [reply xgboost_synthetic/testing/xgboost_test.py]: Good catch; should be setting name.

File: xgboost_synthetic/testing/xgboost_test.py [hunk 2]
```diff
@@ -1,100 +1,125 @@
...
+
+  logging.info("Creating job:\n%s", yaml.dump(job))
+  batch_api.create_namespaced_job(job["metadata"]["namespace"], job)
+  logging.info("Created job %s in namespaces %s", name, namespace)
+
+  # Wait for tiller to be ready
+  end_time = datetime.datetime.now() + datetime.timedelta(
+    minutes=15)
+
+  namespace = job["metadata"]["namespace"]
```
[the remark this answers is not in the document]
Comment (jlewi) [reply xgboost_synthetic/testing/xgboost_test.py]: What's the if you are referring to?
Do you mean setting of name?

File: xgboost_synthetic/testing/xgboost_test.py [hunk 3]
```diff
@@ -1,100 +1,125 @@
...
+    try:
+      job = batch_api.read_namespaced_job(name, namespace)
+    except rest.ApiException as e:
+      logging.error("There was a problem getting job %s.%s; %s",
+                    namespace, name, e)
+      time.sleep(10)
+      continue
+    # ready_replicas could be None
+    if not job.conditions:
+      logging.info("Job missing condition")
```
[the remark this answers is not in the document]
Comment (jlewi) [reply xgboost_synthetic/testing/xgboost_test.py]: Good catch

File: xgboost_synthetic/testing/xgboost_test.py [hunk 4]
```diff
@@ -1,100 +1,125 @@
...
+      break
+    logging.info("Waiting for job %s.%s", namespace, name)
+    time.sleep(10)
+
+  logging.info("Final Job spec:\n%s", yaml.safe_dump(job))
+  util.run(["kubectl", "describe", "job", "-n", namespace, name])
+
+  if not last_condition or last_condition["type"] not in ["Failed", "Complete"]:
+    logging.error("Timeout waiting for job %s.%s to finish.", namespace, name)
+    assert last_condition["type"] in ["Failed", "Complete"]
```
[the remark this answers is not in the document]
Comment (jlewi) [reply xgboost_synthetic/testing/xgboost_test.py]: Done

File: xgboost_synthetic/testing/rolebinding.yaml [hunk 5]
```diff
@@ -1,14 +0,0 @@
-apiVersion: rbac.authorization.k8s.io/v1
```
[the remark this answers is not in the document]
Comment (jlewi) [reply xgboost_synthetic/testing/rolebinding.yaml]: Namespaces should be provisioned with the service account default-editor which should have sufficient priveleges. We shouldn't need to do any additional role creation. If it is then its a bug in Kubeflow somewhere.

File: xgboost_synthetic/testing/xgboost_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jinchihe) [reply xgboost_synthetic/testing/xgboost_test.py]: the ` job["metadata"]["namespace"] = name` above, so no need to `namespace = job["metadata"]["namespace"]`, we can use the namespace var directly. 
Yes I noticed you have updated that. Great!

**Comment by jinchihe** [comment]
Seems there are pylint error in `examples/xgboost_synthetic/util.py `, others LGTM
```
INFO|2019-10-25T00:03:20|util.py:71| ************* Module util
INFO|2019-10-25T00:03:20|util.py:71| C: 15, 0: Line too long (102/100) (line-too-long)
INFO|2019-10-25T00:03:20|util.py:71| C: 37, 0: Trailing whitespace (trailing-whitespace)
INFO|2019-10-25T00:03:20|util.py:71| C: 45, 0: Trailing whitespace (trailing-whitespace)
INFO|2019-10-25T00:03:20|util.py:71| C:  9, 0: standard import "from pathlib import Path" should be placed before "import requests" (wrong-import-order)
```

**Pull request merged**